File: packages/client/src/prediction/commandBuffer.ts
import { createCommand, type Command } from '@ac/shared';

export const COMMAND_BUFFER_CAPACITY = 128;
export const HALF_SEQ_RANGE = 32768;

/** seq 回绕安全的差值：> 0 表示 a 比 b 新。 */
export function seqDiff(a: number, b: number): number {
  const diff = (a - b) & 0xffff;
  return diff >= HALF_SEQ_RANGE ? diff - 65536 : diff;
}

export interface CommandBuffer {
  readonly capacity: number;
  readonly size: number;
  readonly overflowCount: number;
  push(command: Command): void;
  ackUpTo(seq: number): number;
  at(index: number): Command | undefined;
  clear(): void;
}

function copyInto(target: Command, source: Command): void {
  target.seq = source.seq;
  target.tick = source.tick;
  target.moveX = source.moveX;
  target.moveY = source.moveY;
  target.yaw = source.yaw;
  target.pitch = source.pitch;
  target.buttons = source.buttons;
  target.switchTo = source.switchTo;
}

export function createCommandBuffer(capacity: number = COMMAND_BUFFER_CAPACITY): CommandBuffer {
  const cap = Math.max(1, Math.floor(capacity));
  const slots: Command[] = [];
  for (let i = 0; i < cap; i += 1) slots.push(createCommand());
  let head = 0;
  let count = 0;
  let overflowCount = 0;

  return {
    capacity: cap,
    get size(): number {
      return count;
    },
    get overflowCount(): number {
      return overflowCount;
    },
    push(command: Command): void {
      if (count === cap) {
        head = (head + 1) % cap;
        count -= 1;
        overflowCount += 1;
      }
      const slot = slots[(head + count) % cap];
      if (slot === undefined) return;
      copyInto(slot, command);
      count += 1;
    },
    ackUpTo(seq: number): number {
      let dropped = 0;
      while (count > 0) {
        const slot = slots[head];
        if (slot === undefined || seqDiff(seq, slot.seq) < 0) break;
        head = (head + 1) % cap;
        count -= 1;
        dropped += 1;
      }
      return dropped;
    },
    at(index: number): Command | undefined {
      if (index < 0 || index >= count) return undefined;
      return slots[(head + index) % cap];
    },
    clear(): void {
      head = 0;
      count = 0;
    },
  };
}

File: client/Assets/Scripts/Sim/CommandBuffer.cs
namespace Ac.Sim
{
    // C06 §5(a)：未确认命令环形缓冲。容量固定 128，构造期一次分配，Push/AckUpTo 零托管分配。
    public sealed class CommandBuffer
    {
        public const int Capacity = 128;

        private readonly StepCommand[] _slots = new StepCommand[Capacity];
        private int _head;    // 最旧一条
        private int _count;

        public int Size { get { return _count; } }
        public int OverflowCount { get; private set; }

        // §5(a)：ushort 回绕比较；> 0 表示 a 比 b 新。
        public static int SeqDiff(ushort a, ushort b)
        {
            var diff = (a - b) & 0xFFFF;
            return diff >= 32768 ? diff - 65536 : diff;
        }

        public void Push(in StepCommand command)
        {
            if (_count == Capacity)
            {
                _head = (_head + 1) % Capacity;   // 满时丢最旧
                _count -= 1;
                OverflowCount += 1;
            }
            _slots[(_head + _count) % Capacity] = command;
            _count += 1;
        }

        public StepCommand At(int index)
        {
            return _slots[(_head + index) % Capacity];
        }

        // §5(a)：丢弃所有满足 SeqDiff(seq, slot.Seq) >= 0 的最旧条目，返回丢弃条数。
        public int AckUpTo(ushort ackedSeq)
        {
            var dropped = 0;
            while (_count > 0 && SeqDiff(ackedSeq, _slots[_head].Seq) >= 0)
            {
                _head = (_head + 1) % Capacity;
                _count -= 1;
                dropped += 1;
            }
            return dropped;
        }

        public void Reset()
        {
            _head = 0;
            _count = 0;
        }
    }
}
